hstream/src/HStream/Server/Handler.hs

```diff
@@ -361,165 +361,194 @@ sendToClient zkHandle qid sc@SourceConnector {..} ss@streamSend = do
-  let sName = transToStreamName $ TL.toStrict subscriptionStreamName
-  doSubscribe scLDClient reader sName
+
+  isInited <- atomically $ do
+    store <- readTVar subscribeRuntimeInfo
+    if HM.member subscribeRequestSubscriptionId store
+    then return True
+    else return False
+
+  if isInited
+  then returnResp (SubscribeResponse subscribeRequestSubscriptionId)
+  else do
+    -- create a new ldreader for subscription
+    ldreader <- S.newLDRsmCkpReader scLDClient (textToCBytes $ TL.toStrict subscribeRequestSubscriptionId)
+        S.checkpointStoreLogID 5000 1 Nothing 10
+    -- seek ldreader to start offset
+    streamName <- getStreamName subscribeRequestSubscriptionId subscriptions
+    logId <- S.getUnderlyingLogId scLDClient (transToStreamName (TL.toStrict streamName))
+    startOffset <- getSubscriptionOffset subscribeRequestSubscriptionId
+    let startLSN = getStartLSN startOffset
+    S.ckpReaderStartReading ldreader logId startLSN S.LSN_MAX
+    Log.d $ Log.buildString "createSubscribe with startLSN: " <> Log.buildInt startLSN
+    -- insert to runtime info
+    atomically $ do
+      store <- readTVar subscribeRuntimeInfo
+      let newStore = HM.insert subscribeRequestSubscriptionId ldreader store
+      writeTVar subscribeRuntimeInfo newStore
+    -- return resp
+    returnResp (SubscribeResponse subscribeRequestSubscriptionId)
   where
-    -- Confirm subscriptionId exist, and the reader status should be Released, then update the reader status to Occupied.
-    doCheck :: SubscribedReaders -> TL.Text -> IO (S.LDSyncCkpReader, Subscription)
-    doCheck readers subscriptionId = atomically $ do
-      hm <- readTVar readers
-      case HM.lookup subscriptionId hm of
-        Just (ReaderMap rd sId Released) -> do
-          let newMap = HM.insert subscriptionId (ReaderMap rd sId Occupied) hm
-          writeTVar readers newMap
-          return (rd, sId)
-        Just (ReaderMap _ _ Occupied) -> throwSTM SubscriptionIdOccupied
-        Just None -> error "reader status shouldn't be None in subscribeHandler, but get None."
-        Nothing -> throwSTM SubscriptionIdNotFound
-
-    doSubscribe :: S.LDClient -> S.LDSyncCkpReader -> S.StreamId -> IO (ServerResponse 'Normal SubscribeResponse)
-    doSubscribe client reader streamName = do
-      logId <- S.getUnderlyingLogId client streamName
-      S.startReadingFromCheckpoint reader logId S.LSN_MAX
-      currentTime <- getCurrentTimestamp
-      atomically $ modifyTVar' subscribeHeap $ Map.insert subscribeRequestSubscriptionId currentTime
-      returnResp $ SubscribeResponse subscribeRequestSubscriptionId
-
-    doClean :: IO ()
-    doClean = atomically $ updateReaderStatus subscribedReaders Released subscribeRequestSubscriptionId
+    getSubscriptionOffset :: SubscriptionId -> IO SubscriptionOffset
+    getSubscriptionOffset subscriptionId = atomically $ do
+      store <- readTVar subscriptions
+      let Subscription{..} = fromJust $ HM.lookup subscriptionId store
```

Comment:
how about using` case statement` explicit here, we can throw a `SubscriptionIdNotFound` exception if the subscriptionId is not found so that users can know the failed reason more clearly.